### PR TITLE
added banana pi offset mapping

### PR DESCRIPTION
**Pull request by krasi-georgiev**

### CODE REVIEW

**Comment by warthog618** [comment]
The purpose of the Pin function, in the rpi version at least, is to map from string to pin number:

// Pin maps a pin string name to a pin number.

so config read from command line or a file can be readily converted.
Having it map int to int is pointless.

What do the gpiod.go changes have to do with the bananapi?
Not a fan of wrapping errors - it nests badly.  If the user wants to annotate the error by wrapping they are free to, but don't force it on them.
And arbitrarily changing comments?

**Comment by krasi-georgiev** [comment]
The changed comments is just to conform to the golang recommendations
Wrapping the errors is because otherwise there is no way to know where the error came from.
Re the mappings you prefer if I just put the offsets directly for each pin?

**Comment by warthog618** [comment]
Firstly, changes made for distinct reasons should go in distinct commits.  So you've got three or four here, not one.

Generally errors returned by function X don't need to say they were returned by function X - that fact is self evident.  If you want to propagate the error and keep track of the source then you should wrap it yourself.  IMHO anyway.

Wrt the device pin mappings, the point of them is to provide convenience pin name -> number mappings, both in code and from text.  The bananapi looks to be all over the place, even worse than the rpi itself, so I have absolutely no problem with the patch in principle, but the Pin function should be basically the same as the rpi version, minus the j8p section.  For mapping GPIOxx to number in code you already have the GPIO_TO_OFFSET map.  Isn't that sufficient?

File: gpiod.go [hunk 1]
```diff
@@ -8,29 +8,29 @@
...
-//  l, err := gpiod.RequestLine("gpiochip0", 4, gpiod.AsOutput(v))
-//  if err != nil {
-//  	panic(err)
-//  }
-//  for {
-//  	<-time.After(time.Second)
-//  	v ^= 1
-//  	l.SetValue(v)
-//  }
-//
```
Comment (krasi-georgiev) [inline gpiod.go:33]: this is removed from go fmt

**Comment by krasi-georgiev** [comment]
have a look now

**Comment by warthog618** [comment]
That isn't getting closer to what I had in mind.  Or maybe I'm not understanding the preview changes.  
A cut-and-paste of the rpi.Pin, and removing the j8p section was the change to your first commit that I was after (ignoring gpiod.go).

> without errors wrapping and multiple function calls within a function how would you know which function exactly returned the error?

You wrap it where it is called. For each call.  And identify the source as you see fit.  Or provide your own wrapper function that makes the call and wraps any error.  But however you want to do your error handling is application specific - not something that should be embedded in a library.

**Comment by krasi-georgiev** [comment]
> You wrap it where it is called. For each call. And identify the source as you see fit. Or provide your own wrapper function that makes the call and wraps any error. But however you want to do your error handling is application specific - not something that should be embedded in a library.

sorry I don't understand how would that work, can you give an example?

**Comment by warthog618** [comment]
Oh, you mean multiple calls within the library itself.  The calls made within the library is implementation detail, so you really shouldn't be exposing that either.  But generally you can work it out how far the function got from the error type or content.  Got a specific case that is causing you grief?

**Comment by krasi-georgiev** [comment]
yes, was getting some really obscure error  - "no such device or address" and couldn't tell which function returned this. Ended up being an issue with the kernel.

anyway I wouldn't have time to update the PR anytime soon so closing it. Fee free use the mappings if you decide to add support for the bananapi

**Comment by warthog618** [comment]
If the error you are getting an is Errno (in this case you are getting ENXIO) then it came from the kernel, so whichever ioctl call corresponds to your gpiod call.  Errnos normally indicate the line is unavailable for one reason or another.  Generally either you are using the wrong line or there is a driver or board setup issue.

If you are having problems requesting a line you should use the tools to confirm the line is available and that it can be requested.  If it doesn't work with the tools (gpioinfo, gpioget, gpioset,gpiomon,...) it wont work for an app either.  Maybe I should add a troubleshooting section to the docs.

I'm also busy with other things at the moment, but I'll add the bananapi mappings when I get a chance.

**Comment by krasi-georgiev** [comment]
yes you are absolutely right it is a driver's issue
and yes I can match the error in the caller so this approach avoids needing to wrap errors.

https://forum.armbian.com/topic/23799-error-when-trying-to-monitor-gpio-events-with-gpiomon-gpiochip0-40-error-waiting-for-events-no-such-device/?_fromLogin=1

**Pull request closed without merging**